File: src/main/ruby/diffj/ast/imports.rb
#!/usr/bin/jruby -w
# -*- ruby -*-

require 'diffj/ast/element'

module DiffJ
  class ImportsComparator < ElementComparator

    IMPORT_REMOVED = "import removed: {0}"
    IMPORT_ADDED = "import added: {0}"
    IMPORT_SECTION_REMOVED = "import section removed"
    IMPORT_SECTION_ADDED = "import section added"

    def import_to_string imp
      # skip the first token (which is "import")
      str = ""
      tk  = imp.token(0).next
      
      while tk
        if tk == imp.token(-1)
          break
        else
          str << tk.image
          tk = tk.next
        end
      end
      str
    end

    def first_type_token cu
      # this too is a C-style array
      types = cu.type_declarations
      t = types.length > 0 ? types[0].token(0) : nil
      # if there are no types (ie. the file has only a package and/or import
      # statements), then just point to the first token in the compilation unit.      
      t || cu.token(0)
    end

    def mark_import_section_added from, toimports
      fromtk = first_type_token from
      added fromtk, fromtk, toimports[0].token(0), toimports[-1].token(-1), IMPORT_SECTION_ADDED
    end
    
    def mark_import_section_removed fromimports, to
      totk = first_type_token to
      deleted fromimports[0].token(0), fromimports[-1].token(-1), totk, totk, IMPORT_SECTION_REMOVED
    end

    def make_import_map imports
      name_to_imp = Hash.new
      imports.each do |imp|
        str = import_to_string imp
        name_to_imp[str] = imp
      end
      name_to_imp
    end

    def compare_import_blocks fromimports, toimports
      from_names_to_imps = make_import_map fromimports
      to_names_to_imps = make_import_map toimports

      names = from_names_to_imps.keys + to_names_to_imps.keys

      names.each do |name|
        fromimp = from_names_to_imps[name]
        toimp = to_names_to_imps[name]
            
        if fromimp.nil?
          added fromimports[0], toimp, IMPORT_ADDED, name
        elsif toimp.nil?
          deleted fromimp, toimports[0], IMPORT_REMOVED, name
        end
      end
    end

    def compare from, to
      fromimports = from.imports
      toimports = to.imports
      
      if fromimports.empty?
        if !toimports.empty?
          mark_import_section_added from, toimports
        end
      elsif toimports.empty?
        mark_import_section_removed fromimports, to
      else
        compare_import_blocks fromimports, toimports
      end
    end
  end
end

File: src/main/java/org/incava/diffj/compunit/Imports.java
package org.incava.diffj.compunit;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.TreeSet;
import net.sourceforge.pmd.ast.ASTCompilationUnit;
import net.sourceforge.pmd.ast.ASTImportDeclaration;
import net.sourceforge.pmd.ast.ASTTypeDeclaration;
import net.sourceforge.pmd.ast.Token;
import org.incava.diffj.element.Differences;
import org.incava.ijdk.text.Message;
import org.incava.pmdx.CompilationUnitUtil;

public class Imports {
    public static final Message IMPORT_REMOVED = new Message("import removed: {0}");
    public static final Message IMPORT_ADDED = new Message("import added: {0}");
    public static final Message IMPORT_SECTION_REMOVED = new Message("import section removed");
    public static final Message IMPORT_SECTION_ADDED = new Message("import section added");

    private final ASTCompilationUnit compUnit;
    private final List<Import> imports;

    public Imports(ASTCompilationUnit compUnit) {
        this.compUnit = compUnit;
        this.imports = new ArrayList<Import>();
        List<ASTImportDeclaration> astImports = CompilationUnitUtil.getImports(compUnit);
        for (ASTImportDeclaration astImp : astImports) {
            this.imports.add(new Import(astImp));
        }
    }

    public void diff(Imports toImports, Differences differences) {
        if (isEmpty()) {
            if (!toImports.isEmpty()) {
                markAllAdded(toImports, differences);
            }
        }
        else if (toImports.isEmpty()) {
            markAllRemoved(toImports, differences);
        }
        else {
            compareEach(toImports, differences);
        }
    }

    public boolean isEmpty() {
        return imports.isEmpty();
    }

    public ASTImportDeclaration getFirstDeclaration() {
        return imports.get(0).getNode();
    }

    public Token getFirstToken() {
        return imports.get(0).getFirstToken();
    }

    public Token getLastToken() {
        return imports.get(imports.size() - 1).getLastToken();
    }

    public String getImportAsString(ASTImportDeclaration imp) {
        StringBuilder sb = new StringBuilder();   
        Token tk  = imp.getFirstToken().next;
        
        while (tk != null) {
            if (tk == imp.getLastToken()) {
                break;
            }
            else {
                sb.append(tk.image);
                tk = tk.next;
            }
        }

        return sb.toString();
    }

    protected Collection<String> getNames() {
        Collection<String> names = new TreeSet<String>();
        for (Import imp : imports) {
            String str = imp.getAsString();
            names.add(str);
        }
        return names;
    }

    protected ASTImportDeclaration getDeclaration(String name) {
        for (Import imp : imports) {
            if (name.equals(imp.getAsString())) {
                return imp.getNode();
            }
        }
        return null;
    }

    protected void compareEach(Imports toImports, Differences differences) {
        Collection<String> names = getNames();
        names.addAll(toImports.getNames());

        for (String name : names) {
            ASTImportDeclaration fromImp = getDeclaration(name);
            ASTImportDeclaration toImp = toImports.getDeclaration(name);
            
            if (fromImp == null) {
                differences.added(getFirstDeclaration(), toImp, IMPORT_ADDED, name);
            }
            else if (toImp == null) {
                differences.deleted(fromImp, toImports.getFirstDeclaration(), IMPORT_REMOVED, name);
            }
        }
    }

    protected void markAllAdded(Imports toImports, Differences differences) {
        Token fromStart = getFirstTypeToken();
        Token fromEnd = fromStart;
        Token toStart = toImports.getFirstToken();
        Token toEnd = toImports.getLastToken();
        differences.added(fromStart, fromEnd, toStart, toEnd, IMPORT_SECTION_ADDED);
    }

    protected void markAllRemoved(Imports toImports, Differences differences) {
        Token fromStart = getFirstToken();
        Token fromEnd = getLastToken();
        Token toStart = toImports.getFirstTypeToken();
        Token toEnd = toStart;
        differences.deleted(fromStart, fromEnd, toStart, toEnd, IMPORT_SECTION_REMOVED);
    }

    protected Token getFirstTypeToken() {
        List<ASTTypeDeclaration> types = CompilationUnitUtil.getTypeDeclarations(compUnit);
        Token t = types.size() > 0 ? types.get(0).getFirstToken() : null;

        // if there are no types (ie. the file has only a package and/or import
        // statements), then just point to the first token in the compilation
        // unit.

        if (t == null) {
            t = compUnit.getFirstToken();
        }
        return t;
    }
}
